IntelliTect.TestTools.Selenate/IntelliTect.TestTools.Selenate.Tests/ConditionalWaitTests.cs

```diff
@@ -1,32 +1,177 @@
+        }
+
+        [Fact]
+        public async Task CheckFuncParamsForBaseTypeChecking()
+        {
+            await Wait.Until(TestReturnDelegate, TimeSpan.FromSeconds(1), typeof(Exception));
+        }
+
+        [Fact]
+        public async Task CheckActionParamsForMixedTypeChecking()
+        {
+            await Wait.Until(TestVoidDelegate, TimeSpan.FromSeconds(1), typeof(Exception), typeof(NullReferenceException));
+        }
+
+        [Fact]
+        public async Task CheckFuncParamsForMixedTypeChecking()
+        {
+            await Wait.Until(TestReturnDelegate, TimeSpan.FromSeconds(1), typeof(Exception), typeof(NullReferenceException));
+        }
+
+        // Might need explicit scenarios for both func and action scenarios in the below examples.
+        // Gauge based on how much Wait needs to change based on implementation.
+        // Might also need some additional testing around making sure Wait.Until actually times out when expected
+
+        [Fact]
+        public async Task CheckForUnexpectedExceptionType()
+        {
+            await Assert.ThrowsAsync<NullReferenceException>(
+                () => Wait.Until<ArgumentNullException>(
+                    () => ThrowNullRefException(), TimeSpan.FromSeconds(1)));
+        }
+
+        [Fact]
+        public async Task CheckForAggregateExceptionFromExpectedExceptionType()
+        {
+            await Assert.ThrowsAsync<AggregateException>(
+                () => Wait.Until<NullReferenceException>(
+                    () => ThrowNullRefException(), TimeSpan.FromSeconds(1)));
+        }
+
+        [Fact]
+        public async Task CheckForExpectedFailure()
+        {
+            await Assert.ThrowsAsync<AggregateException>(
+                () => Wait.Until<EqualException>(
+                    () => Equals(4, 5), TimeSpan.FromSeconds(1)));
+        }
+
+        [Fact]
+        public async Task CheckForExpectedSuccess()
+        {
+            await Wait.Until<EqualException>(() => Equals(5, 5), TimeSpan.FromSeconds(1));
+        }
+
+        [Fact]
+        public async Task CheckForSuccessAfterFailure()
+        {
+            Stopwatch sw = new Stopwatch();
+            sw.Start();
+            await Wait.Until<InvalidOperationException>(() => IterateResults(2, 1), TimeSpan.FromSeconds(3));
+            sw.Stop();
+            Assert.True(sw.Elapsed < TimeSpan.FromSeconds(3), "Timeout was reached when it wasn't expected to be");
+        }
+
+        [Fact]
+        public async Task CheckForTwoExpectedFailuresByParams()
+        {
+            AggregateException ex = await Assert.ThrowsAsync<AggregateException>(
+                () => Wait.Until(() => IterateResults(5, 2), TimeSpan.FromSeconds(2), typeof(InvalidOperationException), typeof(InvalidProgramException)));
+            Assert.NotNull(ex.InnerExceptions.OfType<InvalidOperationException>());
+            Assert.NotNull(ex.InnerExceptions.OfType<InvalidProgramException>());
```

Comment:
Both of these assertions will always pass since the LINQ `OfType` method will simply return an empty enumerable and never null. I think you probably want to do an `Assert.IsTrue` and use a `.Any()` on the end of those.

Similar comment for the tests below as well.